pkcs11/work.go

```diff
@@ -26,30 +26,62 @@ type Work struct {
 	work *Request // workChan is a channel which has a request enqueue for the worker to work on.
 }
 
-//DoWork performs the work of fetching the signer from the pool and sending it back on the response channel
+// DoWork performs the work of fetching the signer from the pool and sending it back on the response channel.
+// If the client cancels the request or times out, the worker should not wait indefinitely for getting the signer from the
+// pool. Also, we have a max HSM timeout which enforces the worker to wait for max duration to fetch the signer from pool
+// & return if it exceeds that.
 func (w *Work) DoWork(workerCtx context.Context, worker *scheduler.Worker) {
-	select {
-	case <-workerCtx.Done():
-		log.Printf("%s: worker stopped", worker.String())
-		return
-	default:
-		reqCtx, cancel := context.WithTimeout(context.Background(), w.work.remainingTime)
-		defer cancel()
+	reqCtx, cancel := context.WithTimeout(context.Background(), worker.HSMTimeout)
+	type resp struct {
+		signer signerWithSignAlgorithm
+		err    error
+	}
+
+	signerRespCh := make(chan resp)
+	go func() {
 		signer, err := w.work.pool.get(reqCtx)
-		if err != nil {
-			worker.TotalTimeout.Inc()
-			log.Printf("%s: error fetching signer %v", worker.String(), err)
-			w.work.respChan <- nil
-			return
-		}
+		signerRespCh <- resp{signer, err}
```

Comment:
do we need `workerCtx.Done()` check here to ensure this goroutine also shuts down gracefully?